File: src/tests/benchmark/brain.rb
def execute opcodes
	opcode_length = opcodes.length
	memory = []
	loopstack = []
	sp = 0
	ip = 0

	result = ''

	for i in 0...1000
		memory[i] = 0
	end

	while ip < opcode_length
		case opcodes[ip]
			when '+'
				memory[sp] += 1
			when '-'
				memory[sp] -= 1
			when '>'
				sp += 1
			when '<'
				sp -= 1
			when '.'
				result += memory[sp].chr
			when '['
				if memory[sp] != 0
					loopstack.push(ip)
				else
					bracket_matching = 0
					found_index = 0
					(ip..opcode_length).each do |index|
						if opcodes[index] == '['
							bracket_matching += 1
						end
						if opcodes[index] == ']'
							bracket_matching -= 1
						end
						if bracket_matching == 0
							found_index = index
							break
						end
					end

					if bracket_matching != 0
						puts "error: couldn't find match"
						return result
					else
						ip = found_index
					end
				end
			when ']'
				ip = loopstack.pop() - 1
		end
		ip += 1
	end

	result
end

first = execute '++++++++++[>+++++++>++++++++++>+++>+<<<<-]>++.>+.+++++++..+++.>++.<<+++++++++++++++.>.+++.------.--------.>+.'
puts first

second = execute '+++++++++++>+>>>>++++++++++++++++++++++++++++++++++++++++++++>++++++++++++++++++++++++++++++++<<<<<<[>[>>>>>>+>+<<<<<<<-]>>>>>>>[<<<<<<<+>>>>>>>-]<[>++++++++++[-<-[>>+>+<<<-]>>>[<<<+>>>-]+<[>[-]<[-]]>[<<[>>>+<<<-]>>[-]]<<]>>>[>>+>+<<<-]>>>[<<<+>>>-]+<[>[-]<[-]]>[<<+>>[-]]<<<<<<<]>>>>>[++++++++++++++++++++++++++++++++++++++++++++++++.[-]]++++++++++<[->-<]>++++++++++++++++++++++++++++++++++++++++++++++++.[-]<<<<<<<<<<<<[>>>+>+<<<<-]>>>>[<<<<+>>>>-]<-[>>.>.<<<[-]]<<[>>+>+<<<-]>>>[<<<+>>>-]<<[<+>-]>[<+>-]<<<-]'
puts second

File: src/tests/benchmark/brain.py
def execute(opcodes):
	opcode_length = len(opcodes)
	memory = {}
	loopstack = []
	sp = 0
	ip = 0

	result = ''

	for i in range(1000):
		memory[i] = 0

	while ip < opcode_length:
		v = opcodes[ip]
		if v == '+':
				memory[sp] += 1
		elif v == '-':
				memory[sp] -= 1
		elif v == '>':
				sp += 1
		elif v == '<':
				sp -= 1
		elif v == '.':
				result += chr(memory[sp])
		elif v == '[':
				if memory[sp] != 0:
					loopstack.append(ip)
				else:
					bracket_matching = 0
					found_index = 0
					for index in range(ip, opcode_length):
						if opcodes[index] == '[':
							bracket_matching += 1
						if opcodes[index] == ']':
							bracket_matching -= 1
						if bracket_matching == 0:
							found_index = index
							break

					if bracket_matching != 0:
						print("error: couldn't find match")
						return result
					else:
						ip = found_index
		elif v == ']':
				ip = loopstack.pop() - 1
		ip += 1

	return result

first = execute('++++++++++[>+++++++>++++++++++>+++>+<<<<-]>++.>+.+++++++..+++.>++.<<+++++++++++++++.>.+++.------.--------.>+.')
print(first)

second = execute('+++++++++++>+>>>>++++++++++++++++++++++++++++++++++++++++++++>++++++++++++++++++++++++++++++++<<<<<<[>[>>>>>>+>+<<<<<<<-]>>>>>>>[<<<<<<<+>>>>>>>-]<[>++++++++++[-<-[>>+>+<<<-]>>>[<<<+>>>-]+<[>[-]<[-]]>[<<[>>>+<<<-]>>[-]]<<]>>>[>>+>+<<<-]>>>[<<<+>>>-]+<[>[-]<[-]]>[<<+>>[-]]<<<<<<<]>>>>>[++++++++++++++++++++++++++++++++++++++++++++++++.[-]]++++++++++<[->-<]>++++++++++++++++++++++++++++++++++++++++++++++++.[-]<<<<<<<<<<<<[>>>+>+<<<<-]>>>>[<<<<+>>>>-]<-[>>.>.<<<[-]]<<[>>+>+<<<-]>>>[<<<+>>>-]<<[<+>-]>[<+>-]<<<-]')
print(second)

